tidy/impl/test_accurate_type_stubs.py

```diff
@@ -0,0 +1,194 @@
+
+    Returns
+    -------
+    list of str
+        A list of module names from `submod_attrs` that reference `__all__`.
+
+    Notes
+    -----
+    The function uses the `ast` module to parse the source file and identify
+    assignments that contain a `lazy_attach` call with a `submod_attrs` dictionary,
+    where module names are specified with a `__all__` reference.
+    """
+    modules_with_all = []
+    tree = ast.parse(Path(file_path).read_text())
+    for node in tree.body:
+        if (
+            isinstance(node, ast.Assign)
+            and isinstance(call_node := node.value, ast.Call)
+            and isinstance(call_node.func, ast.Name)
+            and call_node.func.id == "lazy_attach"
+        ):
+            for keyword in call_node.keywords:
+                if keyword.arg == "submod_attrs" and isinstance(
+                    keyword.value, ast.Dict
+                ):
+                    submod_attrs_dict = keyword.value
+                    for key_node, value_node in zip(
+                        submod_attrs_dict.keys, submod_attrs_dict.values
+                    ):
+                        if not isinstance(key_node, ast.Constant):
+                            continue
+                        module_name = key_node.value
+                        if (
+                            isinstance(value_node, ast.Attribute)
+                            and value_node.attr == "__all__"
+                            and isinstance(value_node.value, ast.Name)
+                            and module_name == value_node.value.id
+                        ):
+                            modules_with_all.append(module_name)
+    return modules_with_all
+
+
+def get_dunder_all_from_stub(file_path: str) -> List[str]:
+    """
+    Extracts the `__all__` symbols from a type stub (`.pyi`) file.
+
+    Parameters
+    ----------
+    file_path : str
+        The file path of the type stub file to parse.
+
+    Returns
+    -------
+    list of str
+        A list of symbols defined in `__all__` in the type stub.
+
+    Raises
+    ------
+    Exception
+        If `__all__` could not be found in the specified file.
+
+    Notes
+    -----
+    This function specifically looks for assignments to `__all__` in the type
+    stub file and returns a list of the symbols declared in it.
+    """
+    tree = ast.parse(Path(file_path).read_text())
+    for node in tree.body:
+        if isinstance(node, ast.Assign):
+            for target in node.targets:
+                if isinstance(target, ast.Name) and target.id == "__all__":
+                    if isinstance(node.value, ast.List):
+                        return [
+                            elt.s
+                            for elt in node.value.elts
+                            if isinstance(elt, ast.Constant)
+                        ]
+
+    raise Exception("__all__ could not be found")
+
+
+def check_accurate_all_imports(package: List[str]) -> List[str]:
```

Comment:
Could we replace the recursion with os.walk? We would use os.walk to call a function that validates one thing at a time.

Reply:
If we can split this up, I think it'll be a lot easier to tell what's going on

Reply:
Yeah, and it also might be a good idea to split the responsibility of that function into one that makes sure that the __all__ defined in any package is true to that of the type stub, and another that makes sure that the __all__ defined in a subpackage which is imported in some package is also in that packages type stub.